File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupTabLoad.kt
package com.example.momentra.ui.shell.group.shared

import com.example.momentra.data.repository.GroupSliceRepository
import com.example.momentra.ui.shell.perf.ShellPerf
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import java.util.concurrent.ConcurrentHashMap

const val GROUP_PULSE_ACTIVITY_LIMIT = 5

private val pulseInflight =
    ConcurrentHashMap<String, CompletableDeferred<Result<GroupPulseTabData>>>()

/**
 * Paint-first Group pulse load: pulse + finance + activity only.
 * Call [enrichGroupPulseTab] after paint for insights / analytics (non-blocking for spinner).
 * Inflight-deduped so VM prefetch + UI cannot double-fetch.
 */
suspend fun loadGroupPulseTab(
    repository: GroupSliceRepository,
    momentId: String,
    activityLimit: Int = GROUP_PULSE_ACTIVITY_LIMIT,
): Result<GroupPulseTabData> {
    val created = CompletableDeferred<Result<GroupPulseTabData>>()
    val existing = pulseInflight.putIfAbsent(momentId, created)
    val shared = existing ?: created
    if (existing == null) {
        try {
            created.complete(fetchGroupPulseCritical(repository, momentId, activityLimit))
        } catch (t: Throwable) {
            created.complete(Result.failure(t))
        } finally {
            pulseInflight.remove(momentId, created)
        }
    }
    return shared.await()
}

private suspend fun fetchGroupPulseCritical(
    repository: GroupSliceRepository,
    momentId: String,
    activityLimit: Int,
): Result<GroupPulseTabData> = runCatching {
    val mark = ShellPerf.start("pulse_tab_ready")
    coroutineScope {
        val pulseDeferred = async { repository.getPulse(momentId) }
        val financeDeferred = async { repository.getFinance(momentId) }
        val activityDeferred = async { repository.getActivity(momentId, limit = activityLimit) }
        val pulseFacet = pulseDeferred.await().getOrThrow()
        val financeFacet = financeDeferred.await().getOrThrow()
        val activities = activityDeferred.await().getOrThrow().items
        val previous = GroupTabDataCache.peekPulse(momentId)
        val data = GroupPulseTabData(
            title = pulseFacet.title,
            pulse = pulseFacet.payload,
            finance = financeFacet.payload ?: pulseFacet.payload?.finance,
            activities = activities,
            insights = previous?.insights.orEmpty(),
        )
        GroupTabDataCache.putPulse(momentId, data)
        ShellPerf.end(
            mark,
            mapOf("context" to "GROUP", "parallel" to true, "paint" to true, "cached" to false),
        )
        data
    }
}

/**
 * Deferred analytics enrich — does not block Pulse first paint.
 * Merges insights into [GroupTabDataCache] and returns updated tab data.
 */
suspend fun enrichGroupPulseTab(
    repository: GroupSliceRepository,
    momentId: String,
): Result<GroupPulseTabData> = runCatching {
    coroutineScope {
        val insightsDeferred = async { repository.listAnalyticsInsights(scopeId = momentId) }
        val metricsDeferred = async { repository.listAnalyticsMetrics(scopeId = momentId) }
        val refreshDeferred = async { repository.refreshAnalytics(momentId = momentId) }
        val insights = insightsDeferred.await().getOrNull() ?: emptyList()
        metricsDeferred.await()
        refreshDeferred.await()
        val previous = GroupTabDataCache.peekPulse(momentId)
            ?: return@coroutineScope GroupPulseTabData(
                title = null,
                pulse = null,
                finance = null,
                activities = emptyList(),
                insights = insights,
            )
        val data = GroupPulseTabData(
            title = previous.title,
            pulse = previous.pulse,
            finance = previous.finance,
            activities = previous.activities,
            insights = insights,
        )
        GroupTabDataCache.putPulse(momentId, data)
        ShellPerf.instant("group_pulse_enrich", mapOf("momentId" to momentId.take(8)))
        data
    }
}

/**
 * Memory tab load — reuses warm pulse/finance cache when present; only /memory is required.
 * Callers should paint [GroupTabDataCache.peekPulse] / [peekMemory] before awaiting.
 */
suspend fun loadGroupMemoryTab(
    repository: GroupSliceRepository,
    momentId: String,
): Result<GroupTabDataCache.MemoryTab> = runCatching {
    coroutineScope {
        val cached = GroupTabDataCache.peekPulse(momentId)
        val memoryDeferred = async { repository.getMemory(momentId) }
        val financeDeferred = if (cached?.finance != null) null else async { repository.getFinance(momentId) }
        val pulseDeferred = if (cached?.pulse != null) null else async { repository.getPulse(momentId) }
        val participantsDeferred = async { repository.getParticipants(momentId) }
        val memory = memoryDeferred.await().getOrThrow().payload
        val finance = cached?.finance ?: financeDeferred!!.await().getOrThrow().payload
        val pulse = cached?.pulse ?: pulseDeferred!!.await().getOrThrow().payload
        val participants = participantsDeferred.await().getOrNull()?.participants.orEmpty()
        val data = GroupTabDataCache.MemoryTab(
            memory = memory,
            finance = finance,
            pulse = pulse,
            participants = participants,
        )
        GroupTabDataCache.putMemory(momentId, data)
        data
    }
}

/** Prefetch critical-path pulse+finance+activity for Group tab SWR. */
suspend fun prefetchGroupTabs(repository: GroupSliceRepository, momentId: String) {
    if (momentId.isBlank()) return
    loadGroupPulseTab(repository, momentId)
}

File: momentra/momentra/Shell/GroupActive/Shared/GroupTabLoad.swift
import Foundation

/// Paint-first Group pulse load with inflight dedupe (mirrors APK GroupTabLoad).
enum GroupTabLoad {
    private actor InflightGate {
        var tasks: [String: Task<GroupTabDataCache.PulseTab, Error>] = [:]

        func begin(
            momentId: String,
            create: () -> Task<GroupTabDataCache.PulseTab, Error>
        ) -> (task: Task<GroupTabDataCache.PulseTab, Error>, owner: Bool) {
            if let existing = tasks[momentId] { return (existing, false) }
            let task = create()
            tasks[momentId] = task
            return (task, true)
        }

        func end(momentId: String) {
            tasks.removeValue(forKey: momentId)
        }
    }

    private static let activityLimit = GroupTabPrefetch.activityLimit
    private static let gate = InflightGate()

    /// Critical path: pulse + finance + activity. Call [enrich] after paint for analytics.
    static func loadPulseTab(momentId: String) async throws -> GroupTabDataCache.PulseTab {
        let owned = await gate.begin(momentId: momentId) {
            Task { try await fetchCritical(momentId: momentId) }
        }
        do {
            let tab = try await owned.task.value
            if owned.owner {
                await gate.end(momentId: momentId)
            }
            return tab
        } catch {
            if owned.owner {
                await gate.end(momentId: momentId)
            }
            throw error
        }
    }

    private static func fetchCritical(momentId: String) async throws -> GroupTabDataCache.PulseTab {
        let mark = ShellPerf.start("pulse_tab_ready")
        async let pulseResult = APIClient.shared.getGroupPulse(momentId: momentId)
        async let financeResult = APIClient.shared.getGroupFinance(momentId: momentId)
        async let activityResult = APIClient.shared.listGroupActivity(momentId: momentId, limit: activityLimit)
        let loadedPulse = try await pulseResult
        let finFacet = try await financeResult
        let loadedActivity = try await activityResult
        let loadedFinance = finFacet.payload ?? loadedPulse.payload?.finance
        let previous = GroupTabDataCache.peekPulse(momentId)
        let tab = GroupTabDataCache.PulseTab(
            title: loadedPulse.title,
            pulse: loadedPulse,
            finance: loadedFinance,
            activities: loadedActivity,
            insights: previous?.insights ?? []
        )
        GroupTabDataCache.putPulse(momentId, tab)
        ShellPerf.end(mark, extras: ["context": "GROUP", "paint": true, "parallel": true])
        return tab
    }

    /// Deferred analytics — does not block first paint.
    @discardableResult
    static func enrich(momentId: String) async -> GroupTabDataCache.PulseTab? {
        do {
            async let insightsResult = APIClient.shared.listAnalyticsInsights(scopeType: "MOMENT", scopeId: momentId)
            async let metricsResult = APIClient.shared.listAnalyticsMetrics(scopeType: "MOMENT", scopeId: momentId)
            async let refreshResult = APIClient.shared.refreshAnalytics(context: "GROUP_PULSE", momentId: momentId)
            let insights = (try? await insightsResult)?.items ?? []
            _ = try? await metricsResult
            _ = try? await refreshResult
            guard let previous = GroupTabDataCache.peekPulse(momentId) else { return nil }
            let tab = GroupTabDataCache.PulseTab(
                title: previous.title,
                pulse: previous.pulse,
                finance: previous.finance,
                activities: previous.activities,
                insights: insights
            )
            GroupTabDataCache.putPulse(momentId, tab)
            ShellPerf.instant("group_pulse_enrich", extras: [
                "momentId": String(momentId.prefix(8)),
                "insights": insights.count,
            ])
            return tab
        } catch {
            return nil
        }
    }
}
